### PR TITLE
HTTP to HTTP/2 tranlation errors

### PR DESCRIPTION
**Pull request by Scottmitch**
Motivation:
HttpUtil.toHttp2Headers is currently not translating HTTP request headers to HTTP/2 request headers correctly.  The path, scheme, and authority are tranlation process are not respecting the HTTP/2 RFC https://tools.ietf.org/html/rfc7540#section-8.1.2.3 and HTTP RFC https://tools.ietf.org/html/rfc7230#section-5.3.

Modifications:
- path, scheme, authority must be set according to rules defined in https://tools.ietf.org/html/rfc7540#section-8.1.2.3
- HTTP/1.x URIs must be handled as defined in https://tools.ietf.org/html/rfc7230#section-5.3

Result:
More correct translation from HTTP/1.x requests to HTTP/2 requests.

### CODE REVIEW

**Comment by Scottmitch** [comment]
@trustin @nmittler - FYI.

Currently there is an API change for HttpMethod. I can revert if necessary.

StringUtil and other new files will be pushed back to 4.0 unless someone objects.

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpMethod.java [hunk 1]
```diff
@@ -143,18 +133,14 @@ private HttpMethod(String name, boolean bytes) {
...
+        this.name = new AsciiString(name);
+        nameAsString = name;
     }
 
     /**
-     * Returns the name of this getMethod.
+     * Returns the name of this method.
      */
-    public String name() {
+    public AsciiString name() {
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/HttpMethod.java]: @Scottmitch I think we can not do this in 4.1. It's a breaking change. We can only add another method which returns AsciiString.

@trustin agree ?

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpScheme.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+ */
+package io.netty.handler.codec.http;
+
+import io.netty.util.AsciiString;
+
+/**
+ * Defines the common schemes used for the HTTP protocol as defined by
+ * <a href="https://tools.ietf.org/html/rfc7230">rfc7230</a>.
+ */
+public final class HttpScheme {
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/HttpScheme.java]: Seems this is not related to the PR ?

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpMethod.java (hunk 1 above)
Comment (Scottmitch) [inline codec-http/src/main/java/io/netty/handler/codec/http/HttpMethod.java]: http://netty.io/news/2014/12/17/changes-in-deprecation-policy.html

Does deprecating this method in 4.0 allow me to make this change?

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpScheme.java (hunk 2 above)
Comment (Scottmitch) [inline codec-http/src/main/java/io/netty/handler/codec/http/HttpScheme.java]: It is used by this PR. It is easier to leave it in (less PR dependencies, and merge forward to master will be self contained), but I'll separate out the different components of this PR.

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpMethod.java (hunk 1 above)
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/HttpMethod.java]: IHMO no...  Only adding methods would be fine but not change the return value of one

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpScheme.java (hunk 2 above)
Comment (Scottmitch) [inline codec-http/src/main/java/io/netty/handler/codec/http/HttpScheme.java]: See https://github.com/netty/netty/pull/4108.

**Comment by Scottmitch** [comment]
This PR will now fail to build and is dependent on the following PRs: https://github.com/netty/netty/pull/4109 https://github.com/netty/netty/pull/4108 https://github.com/netty/netty/pull/4107

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpMethod.java (hunk 1 above)
Comment (Scottmitch) [inline codec-http/src/main/java/io/netty/handler/codec/http/HttpMethod.java]: It would be nice to have a clarified stance in Netty for the deprecation policy (is it currently defined anywhere).  Not being able to change 4.1 creates a bit of merge/maintenance "hell" (as described in http://netty.io/news/2014/12/17/changes-in-deprecation-policy.html) between 4.0, 4.1, master.  This specific change is small but some others (i.e. headers) may be more complex.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/HttpUtil.java [hunk 3]
```diff
@@ -266,40 +273,58 @@ public static void addHttp2ToHttpHeaders(int streamId, Http2Headers sourceHeader
...
     public static Http2Headers toHttp2Headers(HttpMessage in) throws Exception {
         final Http2Headers out = new DefaultHttp2Headers();
         HttpHeaders inHeaders = in.headers();
         if (in instanceof HttpRequest) {
             HttpRequest request = (HttpRequest) in;
-            out.path(new AsciiString(request.uri()));
-            out.method(new AsciiString(request.method().toString()));
+            /** The path could be absolute-form or authority-form in HTTP, but we should only capture the path
+                component (see <a href="https://tools.ietf.org/html/rfc7230#section-5.3">rfc7230, 5.3</a>). */
+            URI requestTargetUri = URI.create(request.uri());
```
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/HttpUtil.java]: Maybe factor this out into a separate method (e.g. `toHttp2Path(URI)`)?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/HttpUtil.java [hunk 4]
```diff
@@ -266,40 +273,58 @@ public static void addHttp2ToHttpHeaders(int streamId, Http2Headers sourceHeader
...
+            if (!isNullOrEmpty(requestTargetUri.getFragment())) {
+                pathBuilder.append('#');
+                pathBuilder.append(requestTargetUri.getFragment());
+            }
+            String path = pathBuilder.toString();
+            out.path(path.isEmpty() ? EMPTY_REQUEST_PATH : new AsciiString(path));
+            out.method(request.method().name());
 
+            // Attempt to take from HOST header before taking from the request-line
             String value = inHeaders.getAsString(HttpHeaderNames.HOST);
```
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/HttpUtil.java]: maybe rename `value` -> `host`?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/HttpUtil.java (hunk 3 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/HttpUtil.java]: Done

File: codec-http2/src/main/java/io/netty/handler/codec/http2/HttpUtil.java (hunk 4 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/HttpUtil.java]: Done

**Comment by nmittler** [comment]
@Scottmitch LGTM, pending resolution of the deprecation discussion with @normanmaurer

File: codec-http2/src/main/java/io/netty/handler/codec/http2/HttpUtil.java [hunk 5]
```diff
@@ -266,39 +273,41 @@ public static void addHttp2ToHttpHeaders(int streamId, Http2Headers sourceHeader
...
+            if (host == null) {
+                if (shouldSetAuthroity) {
+                    setHttp2Authority(inHeaders, requestTargetUri, out);
                 }
-                value = hostUri.getScheme();
-                if (value != null) {
-                    out.scheme(new AsciiString(value));
+                setHttp2Scheme(inHeaders, requestTargetUri, true, out);
+            } else {
+                URI hostUri = URI.create(host);
```
Comment (trustin) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/HttpUtil.java]: Is this correct? Doesn't the `Host` header have the following format and cannot be parsed by `URI.create()`?

    Host = "Host" ":" host [ ":" port ] ; Section 3.2.2

File: codec-http2/src/test/java/io/netty/handler/codec/http2/HttpToHttp2ConnectionHandlerTest.java [hunk 6]
```diff
@@ -142,6 +144,158 @@ public void testJustHeadersRequest() throws Exception {
...
+                any(ByteBuf.class), anyInt(), anyBoolean());
+    }
+
+    @Test
+    public void testAsterikFormRequestTargetHandled() throws Exception {
+        bootstrapEnv(2, 1, 0);
+        final FullHttpRequest request = new DefaultFullHttpRequest(HTTP_1_1, OPTIONS, "*");
+        final HttpHeaders httpHeaders = request.headers();
+        httpHeaders.setInt(HttpUtil.ExtensionHeaderNames.STREAM_ID.text(), 5);
+        httpHeaders.set(HttpHeaderNames.HOST, "http://www.example.com:80");
```
Comment (trustin) [inline codec-http2/src/test/java/io/netty/handler/codec/http2/HttpToHttp2ConnectionHandlerTest.java]: Shouldn't this be: `www.example.com:80` ?

Comment (trustin) [inline codec-http2/src/test/java/io/netty/handler/codec/http2/HttpToHttp2ConnectionHandlerTest.java]: Nevermind.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/HttpUtil.java (hunk 5 above)
Comment (trustin) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/HttpUtil.java]: Now I see where this is coming from. `URI.create()` cannot parse IPv6 address correctly:

    URI.create("[::1]:80"); // Illegal character in scheme name at index 0: [::1]:80

I wonder if it's illegal to use an IPv6 address in the host part. Obviously, a user can connect using the following URL:

    http://[::1]:80/...

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpMethod.java (hunk 1 above)
Comment (Scottmitch) [inline codec-http/src/main/java/io/netty/handler/codec/http/HttpMethod.java]: @trustin - Thoughts...I'm move vested in the HttpHeaders discussion related to this, but it would be good to know especially before 4.1 is released.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/HttpUtil.java (hunk 5 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/HttpUtil.java]: @trustin - The bracket approach is called out in the standard:

https://tools.ietf.org/html/rfc7230#section-2.7.1 -> https://tools.ietf.org/html/rfc3986#section-3.2.2

```
A host identified by an Internet Protocol literal address, version 6
   [RFC3513] or later, is distinguished by enclosing the IP literal
   within square brackets ("[" and "]").  This is the only place where
   square bracket characters are allowed in the URI syntax.  In
   anticipation of future, as-yet-undefined IP literal address formats,
   an implementation may use an optional version flag to indicate such a
   format explicitly rather than rely on heuristic determination.

      IP-literal = "[" ( IPv6address / IPvFuture  ) "]"

      IPvFuture  = "v" 1*HEXDIG "." 1*( unreserved / sub-delims / ":" )
```

I'll add a unit test for this.

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpMethod.java (hunk 1 above)
Comment (trustin) [inline codec-http/src/main/java/io/netty/handler/codec/http/HttpMethod.java]: @Scottmitch This is an interesting question. I think a user would feel less surprising to remove a deprecated method instead of changing its return type and un-deprecating it.

How about adding `asciiName()` and also keep `name()`? `HttpMethod` is very similar to `Enum`, so it would be less surprising when the return type of `name()` remains as `String`.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/HttpUtil.java (hunk 5 above)
Comment (trustin) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/HttpUtil.java]: Great. Thanks!

Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/HttpUtil.java]: Strictly speaking the scheme is required by https://tools.ietf.org/html/rfc3986#section-3

```
https://tools.ietf.org/html/rfc3986#section-3
```

```java
URI.create("[::1]:80"); // Illegal character in scheme name at index 0: [::1]:80
URI.create("1.2.3.4:80"); // Illegal character in scheme name at index 0: 1.2.3.4:80
URI x = URI.create("localhost:80"); // x.getScheme() == "localhost" and x.getSchemeSpecificPart() == "80"
```

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpMethod.java (hunk 1 above)
Comment (Scottmitch) [inline codec-http/src/main/java/io/netty/handler/codec/http/HttpMethod.java]: @trustin - SGTM.

**Comment by Scottmitch** [comment]
@normanmaurer - Good to cherry-pick?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2CodecUtil.java [hunk 7]
```diff
@@ -261,7 +261,7 @@ public boolean tryFailure(Throwable cause) {
          */
         @Override
         public ChannelPromise setFailure(Throwable cause) {
-            if (awaitingPromises()) {
+            if (awaitingPromises() || expectedCount == 0) {
```
Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2CodecUtil.java]: looks like this is not related to the pr ?

Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2CodecUtil.java]: Yip. Will pull out....unit test in this PR depends upon this though.

Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2CodecUtil.java]: See https://github.com/netty/netty/pull/4123

**Comment by Scottmitch** [comment]
@normanmaurer - After PR https://github.com/netty/netty/pull/4123 ... does this look good to you?

**Comment by normanmaurer** [comment]
@Scottmitch yes

**Comment by Scottmitch** [comment]
4.1 (85c79db) master (0d8ce23)

File: codec-http2/src/main/java/io/netty/handler/codec/http2/HttpUtil.java (hunk 5 above)
Comment (trustin) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/HttpUtil.java]: Huh? Looking from the definition of IP-Literal, isn't `[::1]:80` a valid host string? Given that URI.create() makes us accept invalid Host headers like full URI, shouldn't we use something else?

Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/HttpUtil.java]: > Looking from the definition of IP-Literal, isn't [::1]:80 a valid host string?

Here are the relevant parts of the ABNF:

From https://tools.ietf.org/html/rfc7230:
```
Host = uri-host [ ":" port ] ; Section 2.7.1

uri-host = <host, see [RFC3986], Section 3.2.2>
```

From https://tools.ietf.org/html/rfc3986#section-3.2.2:
```
URI         = scheme ":" hier-part [ "?" query ] [ "#" fragment ]

hier-part   = "//" authority path-abempty
                  / path-absolute
                  / path-rootless
                  / path-empty

authority   = [ userinfo "@" ] host [ ":" port ]

host        = IP-literal / IPv4address / reg-name

IP-literal = "[" ( IPv6address / IPvFuture  ) "]"

IPvFuture  = "v" 1*HEXDIG "." 1*( unreserved / sub-delims / ":" )
```

The way I interpret this is that the scheme is required. Do you agree? Even if so...is this the case in practice...should we be a bit more lenient and permit no scheme'd hosts?

> Given that URI.create() makes us accept invalid Host headers like full URI, shouldn't we use something else?

Can you provide an example? It seems that full URI is what we should expect in the host header?

Comment (trustin) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/HttpUtil.java]: Nope? uri-host is 'host' in 3.2.2 of URI RFC, which is IP literal, IP
address, or regname. It's not a full URI

Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/HttpUtil.java]: @trustin - Agreed.  Sry bout that. Will fix.

Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/HttpUtil.java]: See https://github.com/netty/netty/pull/4132

Comment (trustin) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/HttpUtil.java]: Thank you!
